### DISCUSSION TITLE
How to use latex math environment packages in html?

### DISCUSSION THREAD

**Discussion by fredguth** [quarto-dev/quarto-cli#2976] [Q&A] [answered]
I am building [Tufte-Quarto](https://fredguth.github.io/tufte-quarto/), a Tufte-book.cls based Quarto Project type.  It is a WIP and I still haven't documented it, the site is using my dissertation to showcase/test.

A problem that I face is that I used many packages for the math environment

```latex
\usepackage{bm}
\usepackage{bigints}
\usepackage{nicefrac}
\usepackage{venndiagram}
\usepackage{xpatch}
````
But I don't know how to use them with MathJax in the html render.  I had to comment the code in `Chapters/ai.qmd` that used `\bm` command. 

How do I add these packages to MathJax?

**Comment by mcanouil** [comment]
This is completely a MathJax question which could probably be answered by looking at MathJax documentation: https://docs.mathjax.org/en/latest/
For example, to use "physics" special commands: https://github.com/quarto-dev/quarto-cli/discussions/1620#discussioncomment-3296389
The list of all possible commands and what "package" they come from are listed in the following link: https://docs.mathjax.org/en/latest/input/tex/macros/index.html

**Reply by fredguth** [reply]
...ok, I thought there would be a special metadata for that. A related point, if I may.... 

 Have you noticed that when we add .tex files in the html format, the tex code appears when we change routes?  It flicks when you move from one page to the other.

**Reply by mcanouil** [reply]
No metadata because it's completely dependent of the "math engine", i.e., KaTeX does not use the same logic/syntax.

> Have you noticed that when we add .tex files in the html format, the tex code appears when we change routes? It flicks when you move from one page to the other.

I don't understand what you mean. Do you have some example?

**Reply by fredguth** [reply]
I will try to reproduce and save a video.

Meanwhile, this happens when I render in my machine:
![](https://user-images.githubusercontent.com/392879/197357739-fbb8c88f-33fc-4d51-bc5b-17736f29cc14.png)

But not in gh-pages ¯\_(ツ)_/¯

```_quarto.yml
...
format: # specific to my dissertation
    html:
      html-math-method: mathjax
      include-in-header:
        - text: |
            <script>
              window.MathJax = {
                loader: {
                  load: ['[tex]/bm', '[tex]/bigints', '[tex]/nicefrac', '[tex]/venndiagram', '[tex]/xpatch']
                },
                tex: {
                  packages: {'[+]': ['bm', 'bigints', 'nicefrac', 'venndiagram', 'xpatch']}
                }
              };
            </script>
        - "Tex/_dollar.tex" # hack for macros in html https://github.com/quarto-dev/quarto-cli/discussions/2845
        - "Tex/math_macros.tex"
        - "Tex/_dollar.tex" # hack for macros in html https://github.com/quarto-dev/quarto-cli/discussions/2845
    pdf:
      classoption: ["12pt", "british"]
      include-in-header:
        - Tex/math_commands.tex
        
```
what am I doing wrong here?

**Reply by fredguth** [reply]
Now it is in the `gh-pages` also, I believe it wasn't updated.
[Tufte-Quarto](https://fredguth.github.io/tufte-quarto/)

BTW...
```zsh
 ❯ quarto --version
1.2.215
```

**Reply by fredguth** [reply]
The video of the flickering:
https://share.getcloudapp.com/nOub4z2Q

The text that appears and then disappears is the `Tex/math_macros.tex` file content.

**Comment by fredguth** [comment]
This does not work:
```md
$$
\begin{aligned}
a=1 \tag{TAG}
\\
b=2
\end{aligned}
$$
````

It produces error: `Package amsmath Error: \tag not allowed here.`

I tried:
````
    include-in-header: 
      text: |
        \PassOptionsToPackage{no-math}{fontspec}
        \usepackage[utf8]{inputenc}
        \usepackage{mathtools}  # <--- loads amsmath
````
But still doesn't work.

**Comment by mcanouil** [comment, marked as answer]
It's always better when reporting a bug or that something is not working (as expected) to provide with a fully (and small) reproducible example.

Your last issue seems to be related to #2870

**Reply by fredguth** [reply]
I create a mwe repo here: https://github.com/fredguth/mwe-math-tag

I tried including `ams` as in [#2870], didn't work.

**Reply by fredguth** [reply]
I made a change that made it *kind of* work in html, but errors in pdf.
I pushed the new version to mwe repo.

Here is what I did:
```_quarto.yml

format:
  html:
    theme: cosmo
  pdf:
    documentclass: scrreprt
    include-in-header:
      text: |
          \usepackage{mathtools}
    include-before-body: 
      text: |
          window.MathJax = {
              loader: {
                  load: ['[tex]/tagformat', '[tex]/mathtools', '[tex]/amssymb']
              },
              tex: {
                  packages: {'[+]': ['tagformat', 'mathtools', 'amssymb']},
                  tags: 'ams'
              }
          };
```

``` intro.qmd
# Introduction

## works

$$
\begin{align}
    a=1 \tag{TAG}
    \\
    b=2
\end{align}
$$

```
I changed `aligned` to `align` ... and it is not working in pdf.

**Reply by fredguth** [reply]
I thought MathJax was used only for html, is it been used for pdf render?

**Reply by mcanouil** [reply]
No, it isn't. It's just ignored either by Quarto or Pandoc.
The MathJax part should only be for HTML-based format and be encapsulated in `<script></script>` HTML tags.

`ams` type of tag won't work if you load the `tagformat` extension because it has its own rules.

You are loading a bunch of MathJax extension which changes the way MathJax deals with equations.
Instead, try one at a time, for example, try to use only the `mathtools` tex package and the MathJax equivalent (<https://docs.mathjax.org/en/latest/input/tex/extensions/mathtools.html>), then add the other packages/extensions and configure them.

Also the double dollars environment is a math display environment which is not necessarily needed as you can see in #2870

**Reply by fredguth** [reply]
OMG, your last comment made me grok!
> Also the double dollars environment is a math display environment which is not necessarily needed as you can see in https://github.com/quarto-dev/quarto-cli/issues/2870

`aligned` is for math display environment, and with `align` it is not needed. So, if I just ditched `aligned` for `align` and it works **out of the box** ! (pushed a fix to the mwe repo)

```_quarto.yml
# no change needed
format:
  html:
    theme: cosmo
  pdf:
    documentclass: scrreprt
```
```intro.qmd
# Introduction

## works
\begin{align}
    a=1 \tag{TAG}
    \\
    b=2
\end{align}
```

**Reply by mcanouil** [reply]
The double dollar thing is not trivial.
To see/understand it, the best way is to look at the intermediate markdown/tex files, for example by setting `keep-md: true` and `keep-tex: true`.
